New archival storage for DSN v3. Part 1.

shamil-gadelshin opened the pull request:
This PR is a first step towards the [new design of archival storage (DSN v3)](https://www.notion.so/subspacelabs/DSN-v3-Overview-c50971c4a4f34ca7b0d911a3c202d166#72f4adc1848346afbe747b0a4a2e0603). 

Fixes https://github.com/subspace/subspace/issues/1546 Fixes https://github.com/subspace/subspace/issues/1547
Relates to https://github.com/subspace/subspace/issues/1487

### Changes
The PR adds a new custom protocol for exchanging information about the pieces in farmer's local archival storage. We will use this information when we hit the piece cache(L2) miss. The next PRs will address maintaining connections to a set of farmers to get pieces from, removing the old way of obtaining pieces from L1, removing L1 pieces publication, and adding new method to get pieces from archival storage (L1).

#### Cuckoo-filter
It is used for the exchange of the available pieces in the farmers' archival storage. [The library](https://crates.io/crates/cuckoofilter) is also used in the libp2p project. The default storage capacity is set to 1TB and the custom capacity setting will be addressed in the next part of PRs.
#### Peer-info protocol
We removed the obsolete request-response peer-info protocol and introduced a custom peer-info protocol based on the low-level libp2p primitives (connection handlers, streams, etc). The protocol sends its `peer-info` struct with associated data on the first new connection with peers and also after the "new sector plotted" notification.

#### TODOs
- customizable cuckoo-filter capacity setting
- redesign the peer-info struct
- maintaining persistent connections with a set of farmers

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
farmer: Introduce cuckoo-filter for L1 pieces.

shamil-gadelshin committed:
networking: Remove obsolete peer-info request-response protocol.

shamil-gadelshin committed:
Add peer-info custom protocol.

shamil-gadelshin committed:
Share farmer’s cuckoo filter via peer-info.

shamil-gadelshin committed:
farmer: Notify peers on plotting updates.

nazar-pc commented on crates/subspace-farmer/src/utils/archival_storage_pieces.rs:
```diff
@@ -0,0 +1,38 @@
+use subspace_core_primitives::PieceIndex;
+
+//TODO: make this configurable
+pub const DEFAULT_CAPACITY: usize = (1 << 20) - 1;
```
I don't think we need any default at all. There is no logical default, just remove it and require to specify capacity explicitly.

nazar-pc commented on crates/subspace-farmer/src/utils/archival_storage_pieces.rs:
```diff
@@ -0,0 +1,38 @@
+    pub fn add_pieces(&self, piece_indexes: &[PieceIndex]) -> Result<(), anyhow::Error> {
+        for piece_index in piece_indexes {
+            self.cuckoo_filter
+                .lock()
```
Probably not a good idea performance-wise to take the lock on every iteration of the loop rather than once before the loop

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -354,6 +358,15 @@ where
+                                %disk_farm_index,
+                                %sector_index,
+                                %sector_offset,
+                                "Couldn't add new pieces to archival storage cuckoo filter.");
```
```suggestion
                                "Couldn't add new pieces to archival storage cuckoo filter.",
                            );

nazar-pc commented on crates/subspace-farmer/Cargo.toml:
```diff
@@ -19,6 +19,7 @@ base58 = "0.2.0"
 blake2 = "0.10.6"
 bytesize = "1.2.0"
 clap = { version = "4.2.1", features = ["color", "derive"] }
+cuckoofilter = {version = "0.5.0", features = ["serde_support"] }
```
Attention to details:
```suggestion
cuckoofilter = { version = "0.5.0", features = ["serde_support"] }
```

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+
+/// A [`NetworkBehaviour`] that handles inbound peer info requests and
+/// sends outbound peer info requests on the first established connection.
+pub struct Behaviour<PeerInfoProvider = ConstantPeerInfoProvider> {
```
I would strongly prefer for us to not have default values in such things

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+#[derive(Clone, Debug, Encode, Decode)]
+/// Peer info data
+pub struct PeerInfo {
+    pub role: PeerRole,
```
Public fields should have comments

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+            role: PeerRole::Client,
+        }
+    }
+}
```
`Debug` can be derived, at least for now, if you derive it also on `PeerRole`

shamil-gadelshin replied:
Did you mean `Default` here? `Debug` for `PeerInfo` is customized to avoid printing the whole data vector.

nazar-pc replied:
Yes, `Default`, sorry

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// Handles constant peer info data.
+pub struct ConstantPeerInfoProvider {
```
Wouldn't this evolve over time to include more things? If so why do we need to call it `ConstantPeerInfoProvider`?

shamil-gadelshin replied:
I didn't intend to do that. It was meant for "Client" or "Bootstrap node" peers. It could be renamed though if you suggest better name.

nazar-pc replied:
I'm confused, by the last commits of this PR you have already added `data` field to `PeerInfo` and later updates of cuckoo filter, there doesn't seem to be anything "constant" to it unless I'm missing something.

shamil-gadelshin replied:
Some peers won't announce anything except their assumed roles. They need to return the same data (constant), I could name it `PeerRoleOnlyProvider` instead.

nazar-pc replied:
Hm, I see, you have two implementations. But you configure networking with just one of them, it doesn't seem to be possible to use both at the same time. Why having `ConstantPeerInfoProvider` at all in that case?

shamil-gadelshin replied:
I'm not sure I follow here. Node, BootstrapNode, Client roles use `ConstantPeerInfoProvider` because they won't provider anything except just their role. Farmer has its own implementation. Could you rephrase your question?

nazar-pc replied:
If we encode a message on the wire, decoding side should use exactly the same data structure to be able to decode it. If some nodes on the network use `ConstantPeerInfoProvider` and others use `ArchivalStoragePieces`, how would they even talk to each other? My understanding is that everyone should use the same generics to be compatible.

shamil-gadelshin replied:
Sorry, I still don't see your point. "Message-on-the-wire" is the same for all peers and it is `PeerInfo`, encoding and decoding are also the same and it is SCALE. Different structures provide `PeerInfo` for different peers - that is correct. What issue do you observe?

nazar-pc replied:
I think I now understand what you're saying. Now it seems that `ConstantPeerInfoProvider` is almost identical to `ArchivalStoragePieces` except it is missing said pieces and as the result doesn't fire any events.

What if we have just one `PeerInfoProvider` enum, no traits. It will have several constructors, one for each peer info type. Constructor for farmer will take `ArchivalStoragePieces` as an argument.

That way it is all in one place and `ArchivalStoragePieces` is clearly separated (maybe behind a trait, but probably not necessary initially) as an external source of information.

nazar-pc commented on crates/subspace-farmer/src/utils/archival_storage_pieces.rs:
```diff
@@ -0,0 +1,38 @@
+pub const DEFAULT_CAPACITY: usize = (1 << 20) - 1;
+
+#[derive(Clone)]
+pub struct ArchivalStoragePieces {
```
I'd find a more descriptive name for this, it really not very clear what it is at all. Also everything that is `pub` should have `Debug` implemented.

shamil-gadelshin replied:
`Debug` can't be derived here. I'll add TODO here for renaming. I'm not sure it will be the same at the end of the "new L1 lookups" task. Feel free to suggest options.

nazar-pc replied:
Then implement it manually. We have several examples of this in the codebase, it is not hard to have implementation that at least prints the name of the struct, it doesn't need to do anything fancy.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// The result of an inbound or outbound peer-info requests.
+pub type Result = std::result::Result<PeerInfoSuccess, PeerInfoError>;
```
Such type alias is a bit confusing and misleading. It is also not used very frequently, I see no reason why we should have this type alias, it also causes you to use `std::result::Result` in other places.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+
+            match result {
+                Ok(PeerInfoSuccess::DataSent) => {
+                    debug!(%peer_id,"Peer info sent.", )
```
Here and below:
```suggestion
                    debug!(%peer_id, "Peer info sent.")
```

Also should you print it here or maybe handle/print on the higher level instead?

shamil-gadelshin replied:
It was useful for debugging. If it becomes too noisy I'll remove it or downgrade to the trace level later.

nazar-pc commented on crates/subspace-networking/src/peer_info/handler.rs:
```diff
@@ -0,0 +1,281 @@
+pub enum PeerInfoSuccess {
+    Received(PeerInfo),
+    DataSent,
+}
```
Does it have to be one data structure for both? It seems like we can have separate buffers for these.

Also naming seems off/not symmetrical: if the first variant is `Received` then second must be `Sent` :shrug:
Also comments on enum variants would be helpful (they should have comments, just like any public field of the struct).

shamil-gadelshin replied:
I'm not sure separate structs will have better usage. I renamed options and added comments. Why did `warn(missing_docs)` miss it? Hmm...

nazar-pc replied:
> I'm not sure separate structs will have better usage.

My thinking was that they are not the same kind of message. One is notification and the other is confirmation (right?). So they should not be in the same enum. There could be multiple types of received messages, so one enum variant there would be peer info with potentially more added in the future (let's say incremental updates, so you don't break the protocol every time in incompatible way).

> Why did warn(missing_docs) miss it? Hmm...

I think it might have been a bit confused by the fact that you have defined this struct as `pub`, but that was not correct/necessary because it is not reachable from the outside. You might as well used `pub(crate)`. It maybe it is just a bug in Clippy, those also do happen.

shamil-gadelshin replied:
> My thinking was that they are not the same kind of message. One is notification and the other is confirmation (right?). So they should not be in the same enum.

I don't think I agree with that. 

IMHO they both represent protocol events and can be used for logging, metrics, etc. Some of them are more useful (Received) but they have the same "logical level". Both `ping` and `identify` have the same structure with a minor difference (`identify` adds `Error` to event enum instead of using Result).

nazar-pc commented on crates/subspace-networking/src/peer_info/handler.rs:
```diff
@@ -0,0 +1,281 @@
+#[derive(Debug, Clone)]
+pub struct HandlerInEvent {
+    pub peer_info: PeerInfo,
+}
```
Try not to put structs in between definition and impls of other structs (`Handler` in this case)

nazar-pc commented on crates/subspace-networking/src/peer_info/handler.rs:
```diff
@@ -0,0 +1,281 @@
+            PeerInfoError::Unsupported => None,
+        }
+    }
+}
```
Better use `thiserror`, then you can just write `#[source]` and `Display` will be derived for you

nazar-pc commented on crates/subspace-networking/src/peer_info/handler.rs:
```diff
@@ -0,0 +1,281 @@
+    }
+}
+
+type InPeerInfoFuture = BoxFuture<'static, Result<(NegotiatedSubstream, PeerInfo), io::Error>>;
+type OutPeerInfoFuture = BoxFuture<'static, Result<NegotiatedSubstream, io::Error>>;
```
I wish we didn't box this, maybe add a TODO?

shamil-gadelshin replied:
rust-libp2p uses `BoxFuture` extensively. Introducing special structs and implementing `Future` seems overkill for such cases.

nazar-pc commented on crates/subspace-networking/src/peer_info/handler.rs:
```diff
@@ -0,0 +1,281 @@
+                    }
+                }
+            }
+            Some(OutboundState::Idle(stream)) => {
+                // Nothing to do but we have a negotiated stream.
+                self.outbound = Some(OutboundState::Idle(stream));
+            }
+            Some(OutboundState::NegotiatingStream) => {
+                self.outbound = Some(OutboundState::NegotiatingStream);
+            }
+            Some(OutboundState::RequestNewStream(peer_info)) => {
+                self.outbound = Some(OutboundState::NegotiatingStream);
+                let protocol =
+                    SubstreamProtocol::new(ReadyUpgrade::new(self.config.protocol_name), peer_info)
+                        .with_timeout(self.config.timeout);
+                return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest { protocol });
+            }
```
These for example do not quite look right. The first two do not make any sense, they are no-op and `cx` is not stored, so `poll` will never ever be called again and even if it is by accident. Can you describe why this is the case and how it is supposed to work.

shamil-gadelshin replied:
Those options are not exactly noop, they restore the original `self.outbound` and serve as intermediate states. `Idle` persists the open stream to further reuse and `NegotiatingStream` is just a marker. Both states are changed elsewhere - `Idle` on future peer-info pushes (handler requests) and `NegotiatingStream` in the `FullyNegotiatedOutbound` event handler. Answered about cx and waker in another comment.

nazar-pc commented on crates/subspace-networking/src/peer_info/protocol.rs:
```diff
@@ -0,0 +1,39 @@
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    let mut rec_len_bytes = [0u8; 4];
```
There is nothing wrong here, but I typically prefer writing it in more explicit way:
```
    let mut rec_len_bytes = 0u32.to_le_bytes();
```

Then it will be symmetric to `from_le_bytes()` below.

nazar-pc commented on crates/subspace-farmer/src/utils/archival_storage_pieces.rs:
```diff
@@ -36,3 +36,15 @@ impl ArchivalStoragePieces {
+impl PeerInfoProvider for ArchivalStoragePieces {
+    fn peer_info(&self) -> PeerInfo {
+        let exported_filter = self.cuckoo_filter.lock().export();
+        let data = serde_scale::to_vec(&exported_filter).expect("Serialization always works.");
```
[ExportedCuckooFilter](https://docs.rs/cuckoofilter/latest/cuckoofilter/struct.ExportedCuckooFilter.html) has two public fields, for which you can create a mirror data structure with anything you want derived on it. I don't think it is a better idea to go through `serde_scale` instead.

Also I'm not sure having `data` field as `Vec<u8>` is ideal, I'd rather have the whole `PeerInfo` as an enum, then you'd have `PeerInfo::Farmer {cuckoo_filter: CuckooFilterExported}`, which is nicely serialized/deserialized for you.

And `Debug` will be trivially derivable.

shamil-gadelshin replied:
We use SCALE within the project but most types provide `serde` features. `serde_scale` is a good option IMHO. I don't see downsides except for an additional dependency and we avoid new serialization code. 

`PeerInfo::Farmer {cuckoo_filter: CuckooFilterExported}` - this was my original plan but I'm not sure the type will stay the same in the end. This is why I created a more abstract struct and added TODO both in the code and in the PR description for this case. If the requirements don't change I will change it to the proposed enum. I don't like auto-derived Debug for this type because it prints the whole filter. I added a custom implementation to show only a vector length.

nazar-pc replied:
> We use SCALE within the project but most types provide serde features. serde_scale is a good option IMHO. I don't see downsides except for an additional dependency and we avoid new serialization code.

I'm not exactly sure how `serde_scale` works, but it is surely less efficient than direct implementation. I see no reason to use it if we can have proper and efficient implementation instead. Feels like absolutely unnecessary workaround.

> PeerInfo::Farmer {cuckoo_filter: CuckooFilterExported} - this was my original plan but I'm not sure the type will stay the same in the end. This is why I created a more abstract struct and added TODO both in the code and in the PR description for this case. If the requirements don't change I will change it to the proposed enum.

All of such decisions explicitly do not support upgrades. Why don't you do `PeerInfo::V0(PeerInfoV0)` then with `PeerInfoV0::Farmer {cuckoo_filter: CuckooFilterExported}` so you can always introduce V1 if that is a concern?

I really dislike when we have two field, where one is an enum explaining what the other field contains. You would do that in C because C doesn't support the kinds of enums Rust does, but here it feels like completely unnecessary awkwardness for no benefit.

nazar-pc commented on crates/subspace-farmer/src/utils/archival_storage_pieces.rs:
```diff
@@ -47,4 +56,10 @@ impl PeerInfoProvider for ArchivalStoragePieces {
         }
     }
+
+    fn subscribe(&self, handler: NotificationHandler) -> Option<HandlerId> {
```
Generally convention is to use `fn on_notification` instead of abstract `fn subscribe` and instead of type alias for type alias have a generic `callback` argument here, but for generic you'll probably have to use GATs.

shamil-gadelshin replied:
I renamed the method but I'm not sure about generics here. We have a single type and an alias to hide its complexity. Please, expand here.

nazar-pc replied:
I think it is fine without generic for now, but why did you decide to return an option from here?

shamil-gadelshin replied:
Not all implementations return HandlerId (e.g.: `ConstantPeerInfoProvider`).

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -2,28 +2,54 @@ mod handler;
-#[derive(Clone, Debug, Encode, Decode)]
+#[derive(Debug, Clone, Copy)]
+/// Peer info notification marker.
+pub struct Notification;
```
Why not `()` or even `PeerInfo` rather than require catching useless `Notification` and then calling `.peer_info()` anyway (else why did you subscribe?).

shamil-gadelshin replied:
I'm not sure I follow here. `Notification` is a self-commenting marker. It could be replaced with a type alias of ().

nazar-pc replied:
But you don't care about `Notification`, you care about `PeerInfo`. Why not sending notification with `PeerInfo` instead?

shamil-gadelshin replied:
I tried changing `Notification` to `PeerInfo`. The variant looked more complex to me.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -36,9 +62,13 @@ impl Default for PeerInfo {
     Node,
+    /// DSN bootstrap node.
     BootstrapNode,
+    /// Unspecified client (testing, custom utilities, etc.
```
```suggestion
    /// Unspecified client (testing, custom utilities, etc).
```

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -56,10 +86,17 @@ pub struct Behaviour<PeerInfoProvider = ConstantPeerInfoProvider> {
+    should_notify_handlers: Arc<AtomicBool>,
+    /// We just save the handler ID.
+    #[allow(dead_code)]
+    notify_handler_id: Option<HandlerId>,
```
```suggestion
    _notify_handler_id: Option<HandlerId>,
```

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -150,6 +206,15 @@ impl<PIP: PeerInfoProvider> NetworkBehaviour for Behaviour<PIP> {
+            for peer_id in self.connected_peers.iter().cloned() {
+                self.requests.push(Request { peer_id })
+            }
+        }
```
This doesn't seem quite right either. You check it once, but what if value of `should_notify_handlers` changes to `true`, without waker stored no one will notify futures executor to poll the behavior again. you need to store waker internally and wake it once value changes.

Also `requests` looks more like `messages` or `notifications` (but it would overlap with other notifications) to me. You're not really requesting anything here.

shamil-gadelshin replied:
I adopted "requests-to-handler" concept from `identify` protocol from [libp2p](https://github.com/libp2p/rust-libp2p/blob/8c26c617b94d992c5fbd91efc73049c838740df1/protocols/identify/src/behaviour.rs#L55). Answered about `waker` in another comment.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -0,0 +1,207 @@
+                peer_id,
+                handler: NotifyHandler::Any,
+                event: HandlerInEvent {
+                    peer_info: self.peer_info_provider.peer_info(),
```
`.peer_info()` does expensive locking and serialization that will be exactly the same for everyone, we should ideally call it once and store, maybe in an `Arc` in `Request` itself, so we don't serialize the same thing over and over again for every peer.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -100,11 +144,23 @@ pub struct Event {
+            let should_notify_handlers = should_notify_handlers.clone();
+
+            Arc::new(move |_| {
+                should_notify_handlers.store(true, Ordering::SeqCst);
```
This seems expensive. It would have been MUCH nicer if we sent updates rather than the while filter on every change. Imagine the mess that will happen on the network as any farmers do plotting.

shamil-gadelshin replied:
I'm not sure there are filters supporting it. We send the filter on each sector plotted as we agreed on the DSN v3 design. The alternative would be sending it after we finish the whole plotting (initial suggestion).

nazar-pc replied:
We specifically chose cuckoo filter because it supports updates. Just like filter can be changed locally, it should be possible to change it remotely in incremental fashion, right?

> The alternative would be sending it after we finish the whole plotting (initial suggestion).

It does not fix the problem with replotting, which will happen everywhere all the time and even a single sector update will result in large complete filter being sent.

shamil-gadelshin replied:
It supports updates for the filter but not incremental filter changes unless I miss something.

shamil-gadelshin replied:
> It does not fix the problem with replotting, which will happen everywhere all the time and even a single sector update will result in large complete filter being sent.

Yes, when we implement replotting the current design will send the filter on each sector update. I'm not sure how often we will do that though.

nazar-pc replied:
Half the plot will expire every time history size doubles, so quite often, especially for big plots, especially early during chain history. We need this optimization, we can't just send full filter. If we did and it was infrequent, we'd probably send bloom filters then as they are simpler.

shamil-gadelshin replied:
You have a point. However, our current design doesn't reflect that and I don't have enough information to address the issue. We need to discuss this separately. Here is a dedicated task for it: https://github.com/subspace/subspace/issues/1627

nazar-pc reviewed (commented):
A bunch of nits, questions and some implementation concerns.

shamil-gadelshin committed:
Refactor peer-info protocol.

- change error types
- add comments
- rename methods and types
- add more derived traits
- remove default types
- etc.

shamil-gadelshin reviewed (commented):
It seems that libp2p works without saved wakers ( as recommended, for example, [here](https://tokio.rs/tokio/tutorial/async#wakers). My guess is custom polls in the derived macro's but I couldn't find the exact code. I tested both the initial peer-info exchange and subsequent pushes on the sector plotting - both seem to work as expected. Also, I found a single waker usage in kademlia among libp2p protocols . Please, let me know if I got it wrong.

nazar-pc reviewed (commented):
> It seems that libp2p works without saved wakers ( as recommended, for example, [here](https://tokio.rs/tokio/tutorial/async#wakers). My guess is custom polls in the derived macro's but I couldn't find the exact code. I tested both the initial peer-info exchange and subsequent pushes on the sector plotting - both seem to work as expected. Also, I found a single waker usage in kademlia among libp2p protocols . Please, let me know if I got it wrong.

I'm afraid you did. The fact that it works in libp2p is an accident. There were discussions (including on one of the calls I was on) about fixing it because it is wrong, but some protocols rely on it in some cases and they need to fix it. In short there is ZERO guarantee `poll` will ever be called unless you notify waker (and in non-libp2p case there is close to 100% guarantee it will not be, this is why futures are efficient after all). So IMHO you should not rely on it and it should be fixed to work properly, notifying waker every time you know there is more work to be done in `fn poll`.

nazar-pc commented on crates/subspace-networking/src/peer_info/handler.rs:
```diff
@@ -48,37 +49,30 @@ impl Config {
-            PeerInfoError::Unsupported => None,
-        }
-    }
+/// Marker struct for outbound peer-info requests.
```
Nit about naming: markers don't contain data, they are not a payload themselves either. Markers are usually either traits or structs with no fields used in generics.

shamil-gadelshin committed:
Refactor peer-info protocol.

shamil-gadelshin committed:
networking: Add wakers to peer-info protocol.

shamil-gadelshin committed:
Refactor cuckoo-filter serialization.

shamil-gadelshin committed:
farmer: Calculate cuckoo filter capacity.

shamil-gadelshin reviewed (commented):
I added wakers, please, review the usage among other changes.

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -64,6 +52,32 @@ pub enum PeerRole {
+
+        debug_struct.finish()
+    }
+}
```
Could have been derived, just need to implement `Derive` on `CuckooFilterDTO` manually.

nazar-pc reviewed (commented):
I think there is at least one more waker call missing when `should_notify_handlers` is set to `true`. Otherwise executor will not know that it should send peer info update.

Otherwise this revision looks much better, thanks!

shamil-gadelshin committed:
Refactor peer-info protocol.

shamil-gadelshin committed:
Merge branch 'main' into new-archival-storage3

shamil-gadelshin committed:
Refactor PeerInfoProvider.

shamil-gadelshin committed:
Merge branch 'main' into new-archival-storage3

nazar-pc commented on crates/subspace-networking/src/peer_info.rs:
```diff
@@ -90,34 +91,64 @@ struct Request {
+    /// Creates a new Client peer-info provider.
+    pub fn new_client_provider() -> Self {
+        Self::Client
+    }
```
`_provider` suffix here is unnecessary and not consistent with `fn new_farmer`

shamil-gadelshin committed:
networking: Rename PeerInfoProvider constructors.
